### PR TITLE
[MM-54594] Adding new global relay type

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
We are supporting a new Global Relay export type called "CUSTOM" which allow users set a custom SMTP url and port. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54594

#### Related PRs
https://github.com/mattermost/enterprise/pull/1560

#### Release Note

```release-note
MessageExportSettings.GlobalRelaySettings.CustomerType now supports "CUSTOM".

Added 2 new configs called:
 - MessageExportSettings.GlobalRelaySettings.CustomSMTPServerName
 - MessageExportSettings.GlobalRelaySettings.CustomSMTPPort
 
 They will allow you set a custom URL and port for Global Relay export
 
 CustomSMTPServerName & CustomSMTPPort cannot be configured in the system console UI
```

### CODE REVIEW

File: server/i18n/en.json [hunk 1]
```diff
@@ -8776,7 +8776,11 @@
   },
   {
     "id": "model.config.is_valid.message_export.global_relay.customer_type.app_error",
-    "translation": "Message export GlobalRelaySettings.CustomerType must be set to one of either 'A9' or 'A10'."
+    "translation": "Message export GlobalRelaySettings.CustomerType must be set to one of either 'A9', 'A10' or 'CUSTOM."
```
Comment (hanzei) [inline server/i18n/en.json:8779]: Unrelated comment: Do we support an inbucket export? At least the code does: https://github.com/mattermost/enterprise/blob/589937e32da440afc02a316398a706dd5debba69/message_export/global_relay_export/smtp.go#L40-L56

File: server/public/model/config.go [hunk 2]
```diff
@@ -4076,8 +4085,10 @@ func (s *MessageExportSettings) isValid() *AppError {
...
+			} else if s.GlobalRelaySettings.CustomerType == nil || (*s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA9 && *s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA10 && *s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeCustom) {
 				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.global_relay.customer_type.app_error", nil, "", http.StatusBadRequest)
+			} else if *s.GlobalRelaySettings.CustomerType == GlobalrelayCustomerTypeCustom && (s.GlobalRelaySettings.CustomSMTPServerName == nil || *s.GlobalRelaySettings.CustomSMTPServerName == "") {
+				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.global_relay.customer_type_custom.app_error", nil, "", http.StatusBadRequest)
 			} else if s.GlobalRelaySettings.EmailAddress == nil || !strings.Contains(*s.GlobalRelaySettings.EmailAddress, "@") {
 				// validating email addresses is hard - just make sure it contains an '@' sign
 				// see https://stackoverflow.com/questions/201323/using-a-regular-expression-to-validate-an-email-address
```
Comment (hanzei) [inline server/public/model/config.go:4094]: Unrelated to your changes:
This is really hard to read. Why do we even need all of the  `else if` clauses if the code returns anyway?

File: server/public/model/config.go [hunk 3]
```diff
@@ -4076,8 +4085,10 @@ func (s *MessageExportSettings) isValid() *AppError {
...
 			if s.GlobalRelaySettings == nil {
 				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.global_relay.config_missing.app_error", nil, "", http.StatusBadRequest)
-			} else if s.GlobalRelaySettings.CustomerType == nil || (*s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA9 && *s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA10) {
+			} else if s.GlobalRelaySettings.CustomerType == nil || (*s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA9 && *s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeA10 && *s.GlobalRelaySettings.CustomerType != GlobalrelayCustomerTypeCustom) {
 				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.global_relay.customer_type.app_error", nil, "", http.StatusBadRequest)
+			} else if *s.GlobalRelaySettings.CustomerType == GlobalrelayCustomerTypeCustom && (s.GlobalRelaySettings.CustomSMTPServerName == nil || *s.GlobalRelaySettings.CustomSMTPServerName == "") {
+				return NewAppError("Config.IsValid", "model.config.is_valid.message_export.global_relay.customer_type_custom.app_error", nil, "", http.StatusBadRequest)
```
Comment (hanzei) [inline server/public/model/config.go:4091]: Should the port also get checked?

File: webapp/channels/src/components/admin_console/message_export_settings.jsx [hunk 4]
```diff
@@ -31,7 +31,8 @@ export default class MessageExportSettings extends AdminSettings {
                 SMTPUsername: this.state.globalRelaySMTPUsername,
                 SMTPPassword: this.state.globalRelaySMTPPassword,
                 EmailAddress: this.state.globalRelayEmailAddress,
-
+                CustomSMTPServerName: this.state.globalRelayCustomSMTPServerName,
+                CustomSMTPPort: this.state.globalRelayCustomSMTPPort,
             };
```
Comment (hanzei) [inline webapp/channels/src/components/admin_console/message_export_settings.jsx:36]: Should the values for ServerName and Port be configurable via the system console?

**Review by hanzei — CHANGES_REQUESTED** [review]
The approach is fine, but I'm wondering if we should allow admins to configure the values via the UI.

Heads up that the playwright tests are broken:
`support/server/default_config.ts(624,9): error TS2739: Type '{ CustomerType: string; SMTPUsername: string; SMTPPassword: string; EmailAddress: string; SMTPServerTimeout: number; }' is missing the following properties from type '{ CustomerType: string; SMTPUsername: string; SMTPPassword: string; EmailAddress: string; SMTPServerTimeout: number; CustomSMTPServerName: string; CustomSMTPPort: string; }': CustomSMTPServerName, CustomSMTPPort`

File: webapp/channels/src/components/admin_console/message_export_settings.jsx (hunk 4 above)
Comment (BenCookie95) [reply webapp/channels/src/components/admin_console/message_export_settings.jsx:36]: This is just a quick test for a prospect, we will add that at a later date if our change works for them.

File: server/i18n/en.json (hunk 1 above)
Comment (BenCookie95) [reply server/i18n/en.json:8779]: I think that's only to make some unit tests work?

File: server/public/model/config.go (hunk 3 above)
Comment (BenCookie95) [reply server/public/model/config.go:4091]: Yeah I can add that, I left it out because the port has a default value. No harm in checking anyway

**Review by JulienTant — APPROVED** [review]
Looks good @BenCookie95 thanks

File: server/i18n/en.json (hunk 1 above)
Comment (hanzei) [reply server/i18n/en.json:8779]: Does it actually work in the product if configured via the config file directly?

File: webapp/channels/src/components/admin_console/message_export_settings.jsx (hunk 4 above)
Comment (hanzei) [reply webapp/channels/src/components/admin_console/message_export_settings.jsx:36]: 1/5 that if don't add the UI setting right now, we won't in the future. Is it a substantial amount of work to do that as part of this PR?

File: server/i18n/en.json (hunk 1 above)
Comment (BenCookie95) [reply server/i18n/en.json:8779]: I don't see why it wouldn't work

File: webapp/channels/src/components/admin_console/message_export_settings.jsx (hunk 4 above)
Comment (BenCookie95) [reply webapp/channels/src/components/admin_console/message_export_settings.jsx:36]: Done, it was small

**Review by hanzei — APPROVED** [review]
Awesome, thanks Ben :tada:

**Comment by DHaussermann** [comment]
Hey @BenCookie95 ,
I'm done testing Global Relay for now. 
In terms of regression testing I can't login to the A10 to actually read the data but, that may fall well out of scope as the logging shows me the message count and that it succeeded? 

1. It's weird that it checks if there are any messages to export before seeing if it can authenticate with the config. This made testing a bit confusing at first because it looks like it succeeds when it would not. I can't imagine it's new and I think you even mentioned it yesterday. Maybe I can open a separate enhancement request for this and we can determine the priority as it's somewhat low impact. 

2. The only other thing I have is minor - there aren't user friendly strings to show when the validation fails on save the way existing fields have.  
image

**Comment by BenCookie95** [comment]
> Hey @BenCookie95 , I'm done testing Global Relay for now. In terms of regression testing I can't login to the A10 to actually read the data but, that may fall well out of scope as the logging shows me the message count and that it succeeded?
> 
> 1. It's weird that it checks if there are any messages to export before seeing if it can authenticate with the config. This made testing a bit confusing at first because it looks like it succeeds when it would not. I can't imagine it's new and I think you even mentioned it yesterday. Maybe I can open a separate enhancement request for this and we can determine the priority as it's somewhat low impact.
> 2. The only other thing I have is minor - there aren't user friendly strings to show when the validation fails on save the way existing fields have.
>    image

1. Yeah we can create a separate ticket for that
2. I can make the UI message more user friendly, I'll push something for that in a bit

**Comment by BenCookie95** [comment]
@DHaussermann I can't change that error due to how system console is built. My error format is the same as all the other Global relay errors so I think this is fine

File: server/i18n/en.json (hunk 1 above)
Comment (hanzei) [reply server/i18n/en.json:8851]: @BenCookie95 Would you please open a ticket to add inbucket to the UI, if the code supports it?

**Pull request merged**